### PR TITLE
add test cases wrt DocumentProvider's expected Git interaction

### PR DESCRIPTION
**Pull request by jjelschen**
As a first step, I added a couple of test cases against `DocumentProvider`, which document my assumptions regarding its expected behavior / interaction with Git.

### CODE REVIEW

File: org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/DocumentProviderTest.xtend [hunk 1]
```diff
@@ -18,12 +21,124 @@ class DocumentProviderTest extends AbstractGitTest {
...
+		val filename = "preExistingFile.txt"
+		localGitRoot.newFile(filename).createNewFile
+		git.pull.call
+		git.add.addFilepattern(filename).call
+		git.commit.setMessage("set test preconditions").call
+		git.push.call
+		return filename
+	}
+
+	private def assertSingleCommit(Git git, int numberOfCommitsBefore, ChangeType expectedChangeType, String path) {
```
Comment (gunther-bachmann) [inline org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/DocumentProviderTest.xtend]: suggests that only the expectedChangeType is present, but checks whether at least one is present, right?

File: org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/DocumentProviderTest.xtend [hunk 2]
```diff
@@ -18,12 +21,124 @@ class DocumentProviderTest extends AbstractGitTest {
...
+		val numberOfCommitsBefore = remoteGit.log.call.size
+
+		// when
+		documentProvider.delete(existingFileName)
+
+		// then
+		remoteGit.assertSingleCommit(numberOfCommitsBefore, DELETE, existingFileName)
+	}
+
+	private def preExistingFileInRepo() {
```
Comment (gunther-bachmann) [inline org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/DocumentProviderTest.xtend]: how about naming it more like 'create and push file to repo'

File: org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/DocumentProviderTest.xtend [hunk 3]
```diff
@@ -18,12 +21,124 @@ class DocumentProviderTest extends AbstractGitTest {
...
+	}
+
+	@Test
+	def void savePushesChanges() {
+		// given
+		val existingFileName = preExistingFileInRepo
+		val numberOfCommitsBefore = remoteGit.log.call.size
+
+		// when
+		documentProvider.save(existingFileName, "Contents of pre-existing file")
```
Comment (gunther-bachmann) [inline org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/DocumentProviderTest.xtend]: should be "new contents for existing file" (as you did in createOrUpdatePreExistingFileCommitsModifications)

File: org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/DocumentProviderTest.xtend [hunk 4]
```diff
@@ -18,12 +21,124 @@ class DocumentProviderTest extends AbstractGitTest {
...
+		val numberOfCommitsBefore = git.log.call.size
+
 		// when
 		documentProvider.create('a.txt', 'test')
 
 		// then
-		val commits = git.log.call
-		commits.assertSize(2)
+		val numberOfCommitsAfter = git.log.call.size
+		numberOfCommitsAfter.assertEquals(numberOfCommitsBefore + 1)
```
Comment (gunther-bachmann) [inline org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/DocumentProviderTest.xtend]: why not use the assertSingleCommit function you use in other test methods?

File: org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/git/AbstractGitTest.xtend [hunk 5]
```diff
@@ -27,6 +34,8 @@ abstract class AbstractGitTest extends AbstractPersistenceTest {
 
 	@Inject protected GitProvider gitProvider
 	@Mock protected WorkspaceProvider workspaceProvider
+	
+	protected Git remoteGit
```
Comment (gunther-bachmann) [inline org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/git/AbstractGitTest.xtend:38]: should (only) be accessible via getter, right?

File: org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/git/AbstractGitTest.xtend [hunk 6]
```diff
@@ -56,5 +65,31 @@ abstract class AbstractGitTest extends AbstractPersistenceTest {
...
+		val commit = walk.parseCommit(lastCommitId)
+		return commit
+	}
+	
+	
+	/**
+	 * Helper method for calculating the diff of a Git commit.
+	 */
+	protected def List<DiffEntry> getDiffEntries(Git git, RevCommit commit) {
+		val repository = git.repository
```
Comment (gunther-bachmann) [inline org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/git/AbstractGitTest.xtend:82]: MINOR: if you drop this line, you could leave out the df binding in line 87, setting repository = git.repository

File: org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/DocumentProviderTest.xtend (hunk 4 above)
Comment (jjelschen) [reply org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/DocumentProviderTest.xtend]: This was the only test case that was already present (I just modified it), and somehow I must have missed it when I introduced the helper method during a later refactoring.

File: org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/DocumentProviderTest.xtend (hunk 3 above)
Comment (jjelschen) [reply org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/DocumentProviderTest.xtend]: Right, applies for both `saveCommitsChanges()` and `savePushesChanges()`.

File: org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/DocumentProviderTest.xtend (hunk 1 above)
Comment (jjelschen) [reply org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/DocumentProviderTest.xtend]: hm, the "single" refers to a single commit, not a single diff entry. In my current test cases, I always have only a single diff entry, too, but you're right, it checks if there is at least one diff entry that satisfies the condition.

So… `assertSingleCommitContaining`, and then maybe reorder the arguments so it reads better? The `numberOfCommitsBefore` is still the odd one out, though.

**Pull request merged**